pkg/resto/resto.go

```
@@ -0,0 +1,357 @@
+			Size:     int64(len(v)),
+		}); err != nil {
+			return nil, "", err
+		}
+		if _, err := w.Write([]byte(v)); err != nil {
+			return nil, "", err
+		}
+	}
+	w.Close()
+	dgst := digest.SHA256.FromBytes(buf.Bytes())
+	gzbuf := bytes.NewBuffer(nil)
+	g := gzip.NewWriter(gzbuf)
+	g.Write(buf.Bytes())
+	g.Close()
+	return gzbuf.Bytes(), dgst, nil
+}
+
+// ListRegistry lists all the repositories in a registry
+func ListRegistry(ctx context.Context, endpoint, username, password string, insecure bool) ([]string, error) {
+	tr, err := NewTransportCatalog(endpoint, username, password, insecure)
+	if err != nil {
+		return nil, err
+	}
+	registry, err := client.NewRegistry(endpoint, tr)
+	if err != nil {
+		return nil, err
+	}
+	entries := make([]string, 10000)
+	count, err := registry.Repositories(ctx, entries, "")
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+	return entries[0:count], nil
+}
+
+// ListRepository lists all the tags in a repository
+func ListRepository(ctx context.Context, reponame, username, password string, insecure bool) ([]string, error) {
+	pr, err := parseRef(reponame)
+	if err != nil {
+		return nil, err
+	}
+	repo, err := NewRepository(ctx, pr.domain, pr.path, username, password, insecure)
+	if err != nil {
+		return nil, err
+	}
+	tagService := repo.Tags(ctx)
+	return tagService.All(ctx)
+}
+
+// PullConfig pulls a configuration file from a registry
+func PullConfig(ctx context.Context, repoTag, username, password string, insecure bool) (string, error) {
+	res, err := PullConfigMulti(ctx, repoTag, username, password, insecure)
+	if err != nil {
+		return "", err
+	}
+	return res["config"], nil
+}
+
+// PullConfigMulti pulls a set of configuration files from a registry
+func PullConfigMulti(ctx context.Context, repoTag, username, password string, insecure bool) (map[string]string, error) {
+	pr, err := parseRef(repoTag)
+	if err != nil {
+		return nil, err
+	}
+	if username == "" {
+		username, password, _ = getCredentials(pr.domain)
```

Why ignoring the error here ?

---

Because failure to obtain credentials is not an error. One might not need them if pushing to a plain registry.